aggie-coding-club/Rev-Registration
# Rename schedules

## Description

Reorganizes the schedules redux (and creates a schedule type so that it can be more flexible in the future), and adds the ability to rename schedules.

Draft for now since I want `frontend/delete-schedule-dialog` merged before this gets worked on more, although it's completely functional with tests (let me know if there are any other tests I should add though)

## Rationale

I made a `Schedule` type since each schedule now needs a name instead of just a list of meetings, which is a lot more flexible than how it was being done before but involved refactoring a decent amount of old code. This makes some parts of our old code a little confusing, like how some meetings arrays were referred to as a schedule in the past, but I opted not to change these for now.

The behavior for creating schedules with identical names could also be a bit controversial? Here are some examples:

Schedule with name `test` exists -> rename another schedule to `test` -> new `test` schedule is renamed to `test (2)`
Schedule with name `Schedule 1` exists -> rename another schedule to `Schedule 1` -> new schedule is renamed to `Schedule 1 (2)`
Schedule with name `Schedule 1` is saved -> generate more schedules -> new schedules generated are `Schedule 2`, `Schedule 3`, etc.

## How to test

Click the rename button and try renaming some schedules/generating more to see how it works (also try creating 2 schedules with the same name and saving a schedule with the default name).

Also try using keyboard controls, I made it autofocus on the schedule name when you click edit so you don't have to click the name input, and you can cancel renaming with escape or confirm with enter.

## Screenshots
|Before|After|
|--|--|
| ![image](https://user-images.githubusercontent.com/28655462/92332434-b77b2f00-f043-11ea-8a6a-dee9af9dae04.png) | ![image](https://user-images.githubusercontent.com/28655462/92332420-a8947c80-f043-11ea-900c-60682c739644.png) |
| | ![image](https://user-images.githubusercontent.com/28655462/92332467-c82ba500-f043-11ea-909e-fda09d5ff292.png) |

## Related tasks

Closes #296.
